Code/GraphMol/FileParsers/GeneralFileReader.h

```diff
@@ -0,0 +1,233 @@
+
+    //! if the compression type is not valid then we could be in the case
+    //! where the file name is of the form *.2.txt and so set the file format
+    //! (which was earlier .2) as the compression format (which was earlier
+    //! .txt) and set the compression format as an empty string
+    if (!flag_compressionFormat) {
+      fileFormat = compressionFormat;
+      compressionFormat = "";
+      flag_fileFormat = std::find(fileFormats.begin(), fileFormats.end(),
+                                  fileFormat) != fileFormats.end();
+    }
+  }
+  return flag_fileFormat;
+}
+
+//! returns the file name givens the file path
+std::string getFileName(const std::string path) {
+  char delimiter = '/';
+  std::string fname = "";
+  auto slash = path.rfind(delimiter, path.length());
+  if (slash != std::string::npos) {
+    fname += path.substr(slash + 1, path.length() - slash);
+  }
+  return fname;
+}
+
+//! given file path determines the file and compression format
+void determineFormat(const std::string path, std::string& fileFormat,
+                     std::string& compressionFormat) {
+  std::string fileName = getFileName(path);
+  int dots = std::count(fileName.begin(), fileName.end(), '.');
+
+  if (dots == 0)
+    throw std::invalid_argument(
+        "Recieved Invalid File Format, no extension or compression");
+
+  else if (dots == 1) {
+    //! there is a file format but no compression format
+    int pos = fileName.find(".");
+    fileFormat = fileName.substr(pos + 1);
+    if (!valid(fileFormat, compressionFormat))
+      throw std::invalid_argument("Recieved Invalid File Format");
+  } else {
+    //! there is a file and compression format
+    int n = fileName.length();
+    int p1 = fileName.rfind(".");
+    int p2 = fileName.rfind(".", p1 - 1);
+    fileFormat = fileName.substr(p2 + 1, (p1 - p2) - 1);
+    //! possible compression format
+    compressionFormat = fileName.substr(p1 + 1, (n - p1) + 1);
+    if (!valid(fileFormat, compressionFormat))
+      throw std::invalid_argument(
+          "Recieved Invalid File or Compression Format");
+  }
+}
+
+//! returns a MolSupplier object based on the file name instantiated
+//! with the relevant options provided in the SupplierOptions struct
+/*!
+    <b>Note:</b>
+      - the caller owns the memory and therefore the pointer must be deleted
+*/
+MolSupplier* getSupplier(const std::string& path,
+                         const struct SupplierOptions opt) {
+  std::string fileFormat = "";
+  std::string compressionFormat = "";
+  //! get the file and compression format form the path
+  determineFormat(path, fileFormat, compressionFormat);
+
+  //! Handle the case when there is no compression format
+  if (compressionFormat.empty()) {
+    std::ifstream* strm = new std::ifstream(path.c_str());
+    if (fileFormat.compare("sdf") == 0) {
+      ForwardSDMolSupplier* sdsup =
+          new ForwardSDMolSupplier(strm, opt.takeOwnership, opt.sanitize,
```

Comment:
you always need the suppliers to take ownership of the stream, otherwise the memory will be leaked. So instead of using `opt.takeOwnership` here, I would suggest just passing `true` as that argument.
This comment applies to all the places where you create a supplier